Rename InternedString to LocalInternedString and introduce a new thread-safe InternedString

Zoxc opened the pull request:
This is an allocation-free alternative to https://github.com/rust-lang/rust/pull/46972.

cc @rust-lang/compiler 

r? @michaelwoerister

rust-highfive commented:
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/365408340) ([raw log](https://api.travis-ci.org/v3/job/365408340/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:00:46] configure: rust.quiet-tests     := True
---
[00:05:53] tidy error: /checkout/src/libsyntax_pos/symbol.rs:399: line longer than 100 chars
[00:05:54] some tidy checks failed
[00:05:54]
[00:05:54]
[00:05:54] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor" "--quiet"
[00:05:54] expected success, got: exit code: 1
[00:05:54]
[00:05:54]
[00:05:54] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:05:54] Build completed unsuccessfully in 0:02:30
[00:05:54] Makefile:79: recipe for target 'tidy' failed
[00:05:54] make: *** [tidy] Error 1
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/365537678) ([raw log](https://api.travis-ci.org/v3/job/365537678/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
Resolving deltas: 100% (614309/614309), completed with 4928 local objects.
---
[00:00:56] configure: rust.quiet-tests     := True
---
[00:33:14] 1362 |         let path = external_path(cx, &cx.tcx.item_name(did),
[00:33:14]      |                                      ^^^^^^^^^^^^^^^^^^^^^^ expected str, found struct `syntax::symbol::InternedString`
---
[00:33:14] 1469 |         let path = external_path(cx, &cx.tcx.item_name(trait_ref.def_id),
[00:33:14]      |                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected str, found struct `syntax::symbol::InternedString`
---
[00:33:16] 2796 |                 let path = external_path(cx, &cx.tcx.item_name(did),
[00:33:16]      |                                              ^^^^^^^^^^^^^^^^^^^^^^ expected str, found struct `syntax::symbol::InternedString`
---
[00:33:16] 2807 |                 let path = external_path(cx, &cx.tcx.item_name(did),
[00:33:16]      |                                              ^^^^^^^^^^^^^^^^^^^^^^ expected str, found struct `syntax::symbol::InternedString`
---
[00:33:16] 2825 |                         let path = external_path(cx, &cx.tcx.item_name(did),
[00:33:16]      |                                                      ^^^^^^^^^^^^^^^^^^^^^^ expected str, found struct `syntax::symbol::InternedString`
---
[00:33:17] 2848 |                     let path = external_path(cx, &cx.tcx.item_name(did), Some(did),
[00:33:17]      |                                                  ^^^^^^^^^^^^^^^^^^^^^^ expected str, found struct `syntax::symbol::InternedString`
---
[00:33:19] 264 |                         hir::PathSegment::from_name(Symbol::intern(&param.name))
[00:33:19]     |                                                                    ^^^^^^^^^^^ expected str, found struct `syntax::symbol::InternedString`
---
[00:33:25] Makefile:28: recipe for target 'all' failed
[00:33:25] make: *** [all] Error 1
---
$ ls -lat $HOME/Library/Logs/DiagnosticReports/
ls: cannot access /home/travis/Library/Logs/DiagnosticReports/: No such file or directory
travis_time:end:008b7d54:start=1523530164931732610,finish=1523530164944951894,duration=13219284
travis_fold:end:after_failure.2
travis_fold:start:after_failure.3
travis_time:start:1122f171
$ find $HOME/Library/Logs/DiagnosticReports -type f -name '*.crash' -not -name '*.stage2-*.crash' -not -name 'com.apple.CoreSimulator.CoreSimulatorService-*.crash' -exec printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" {} \; -exec head -750 {} \; -exec echo travis_fold":"end:crashlog \; || true
find: `/home/travis/Library/Logs/DiagnosticReports': No such file or directory
travis_time:end:1122f171:start=1523530164956619488,finish=1523530164969050438,duration=12430950
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:083dd5fa
$ dmesg | grep -i kill
[   10.852881] init: failsafe main process (1092) killed by TERM signal
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

petrochenkov commented:
Why this distinction is necessary? When one should use `InternedString` and when `LocalInternedString`? I can't infer it from this patch (I haven't followed other parallelization PRs though).

I planned to get rid of `InternedString` completely in favor of `Symbol` after https://github.com/rust-lang/rust/issues/49300 is fixed (and `StableInternedString` -> `StableSymbol` for string-y `Ord` and stuff).

Zoxc commented:
> Why this distinction is necessary?

`LocalInternedString` has a `Deref<Target=str>` impl which makes it more expressive to use, but also makes it not thread-safe (it doesn't implement `Send`). `InternedString` has no `Deref<Target=str>` impl and implements `Send` and `Sync` which makes it suitable to be shared between multiple threads.

> When one should use InternedString and when LocalInternedString?

Pick `InternedString` if you are sharing or sending things between threads (the compiler will inform you). Otherwise you can pick either one.

> I planned to get rid of InternedString completely in favor of Symbol

You can still get rid of the introduced `InternedString` in favor of Symbol. It still makes sense to keep around `LocalInternedString` because of its more expressive API, but it should probably be renamed to `LocalSymbol` at that point.

petrochenkov commented:
I've just noticed that string interner is no longer thread local and every access through `Symbol`'s index into interner's vector goes through a lock.
So it looks like `LocalInternedString` in addition to being a convenience also serves as an optimization.
For convenience alone I'd personally wouldn't add an extra string type, but with locks I guess there's no way to avoid this string type proliferation. Sigh.

Zoxc commented:
@petrochenkov We can replace everything with `Symbol(&'tcx str)` once querification reaches libsyntax =P

petrochenkov commented:
>interner is no longer thread local and every access through Symbol's index into interner's vector goes through a lock

Actually, now I understand even less. If the interner is global, then string contents live effectively forever once interned and can be accessed from all threads, then why `Deref` to `str` is incompatible with `Sync`/`Send`. Please, ELI5 how this works.

Mark-Simulacrum commented:
I'm also somewhat confused in general about the design of `Interner`; it feels like we could relatively easily get away with something like what is implemented in this [repo](https://github.com/Mark-Simulacrum/string-internment/blob/master/src/lib.rs) -- the API isn't one-to-one compatible, but the potential API should be effectively the same as today's Symbol/Interner API, if not easier to use. The approach is one where we basically just leak the data (which must be true for today's implementation of `Interner` if it's safe, since we have safe, public `as_str` on symbol that returns `&'static str`.

The negative there is that the size of `Symbol` increases from 4 bytes to 16 bytes as it goes from `u32` to `&'static str`. I don't know what the size and performance impact of that would be.

----

This might be best to discuss elsewhere, but it's also confusing to me why we make `InternedString` not implement `Send` when it does implement `Sync`. Since it's a `Copy` type that *seems* to imply that not implementing Send means nothing -- I can share a reference between two threads and then copy it in both threads. (Maybe this means a lint on this situation would be good?)

bors commented:
:umbrella: The latest upstream changes (presumably #49718) made this pull request unmergeable. Please resolve the merge conflicts.

Zoxc commented:
> The approach is one where we basically just leak the data (which must be true for today's implementation of Interner if it's safe, since we have safe, public as_str on symbol that returns &'static str.

We do not want to leak strings. The idea of `LocalInternedString` is that it contains a `&'thread str` where `thread` is the lifetime of the current thread. So in order to ensure `LocalInternedString` is sound we just need to have the interner (which contains the `str`) outlive threads we create `LocalInternedString`s on.

> This might be best to discuss elsewhere, but it's also confusing to me why we make InternedString not implement Send when it does implement Sync. Since it's a Copy type that seems to imply that not implementing Send means nothing -- I can share a reference between two threads and then copy it in both threads. (Maybe this means a lint on this situation would be good?)

`LocalInternedString` can't implement both `Sync` and `Copy`, so I added a commit to remove the Sync impl.

Mark-Simulacrum commented:
> We do not want to leak strings.

Yet isn't that basically a requirement for today's implementation?

"This accesses the thread-local interner in order to get to the string." is the comment on `InternedString` which can safely be retrieved from `LocalInternedString`. I also see "This is safe because the interner never frees strings" in the code; that also implies that we're effectively leaking strings today while still paying the overhead for keeping `InternedString` not just pointing to the string data, and instead having to go through Symbol to get to the data. Am I missing something? What is the difference between `LocalInternedString` and `InternedString`?

It seems to me that long-term `Symbol` should become `&'static str` (perhaps wrapped in a newtype) and that should be feasible and is *almost* what we have today anyway.

rust-highfive commented:
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/366805780) ([raw log](https://api.travis-ci.org/v3/job/366805780/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:00:47] configure: rust.quiet-tests     := True
---
serialize-c04ded78717d5d67.rlib --extern syntax_pos=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libsyntax_pos-9f3518d56a01456f.so --extern fmt_macros=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libfmt_macros-3706e912fdb98df1.so --extern bitflags=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libbitflags-9866e194db82a141.rlib --extern graphviz=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libgraphviz-0bde40de32995f14.so --extern flate2=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libflate2-3c8223b0152f22a5.rlib -L native=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/build/backtrace-sys-8b35e3c2ea935fab/out/.libs -L native=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/build/miniz-sys-63734d0048644b22/out` (exit code: 101)
[00:07:49] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "--target" "x86_64-unknown-linux-gnu" "--release" "--locked" "--color" "always" "--features" " jemalloc" "--manifest-path" "/checkout/src/rustc/Cargo.toml" "--message-format" "json"
[00:07:49] expected success, got: exit code: 101
[00:07:49] thread 'main' panicked at 'cargo must succeed', bootstrap/compile.rs:1085:9
[00:07:49] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:07:49] travis_fold:end:stage0-rustc
[00:07:49] travis_time:end:stage0-rustc:start=1523804678821502182,finish=1523804855296100416,duration=176474598234
[00:07:49] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap build
[00:07:49] Build completed unsuccessfully in 0:03:08
[00:07:49] Makefile:28: recipe for target 'all' failed
[00:07:49] make: *** [all] Error 1
---
$ find $HOME/Library/Logs/DiagnosticReports -type f -name '*.crash' -not -name '*.stage2-*.crash' -not -name 'com.apple.CoreSimulator.CoreSimulatorService-*.crash' -exec printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" {} \; -exec head -750 {} \; -exec echo travis_fold":"end:crashlog \; || true
find: `/home/travis/Library/Logs/DiagnosticReports': No such file or directory
travis_time:end:151da37b:start=1523804855738312853,finish=1523804855744662139,duration=6349286
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:173af718
$ dmesg | grep -i kill
[   10.535064] init: failsafe main process (1094) killed by TERM signal
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Zoxc commented:
> Yet isn't that basically a requirement for today's implementation?

No. The requirement is that the interner outlives any threads using `LocalInternedString`. Since `LocalInternedString` is not `Send` nor `Sync` and the `&'static str` field is private, there is no way for the string reference to escape the thread using it.

> I also see "This is safe because the interner never frees strings" in the code

It never frees strings until it is dropped, then all the strings are freed. I changed this comment to make it clearer.

> "This accesses the thread-local interner in order to get to the string." is the comment on InternedString which can safely be retrieved from LocalInternedString

I'm not sure what your point is there, but the comment is incorrect since we the interner is no longer thread-local. I removed that.

rust-highfive commented:
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/366807705) ([raw log](https://api.travis-ci.org/v3/job/366807705/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:00:48] configure: rust.quiet-tests     := True
---
--extern rustc_errors=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_errors-609e2421d03f9c9a.so --extern rustc_apfloat=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_apfloat-1299638d641ea770.rlib --extern rustc_back=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_back-861dba9fe03aa669.so --extern graphviz=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libgraphviz-0bde40de32995f14.so --extern jobserver=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libjobserver-5bcc8c1ccd509892.rlib --extern lazy_static=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/liblazy_static-2803512e03c19cf7.rlib --extern proc_macro=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libproc_macro-6e4119b5ec8457a3.so --extern rustc_const_math=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_const_math-895f7ddc4467bb8d.so --extern tempdir=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libtempdir-65dde0349e75e965.rlib --extern rustc_data_structures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_data_structures-93cb1ddd29ab61a4.so --extern fmt_macros=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/releasestage0-rustc:start=1523805119657710455,finish=1523805319582979368,duration=199925268913
[00:08:16] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap build
[00:08:16] Build completed unsuccessfully in 0:03:33
[00:08:16] Makefile:28: recipe for target 'all' failed
[00:08:16] make: *** [all] Error 1
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Mark-Simulacrum commented:
Ah, I think I understand now. I believe the root of my confusion came from the fact that without parallel queries being enabled, neither `InternedString` nor `LocalInternedString` are `Sync/Send`, which makes the API this PR proposes fine.

In order for this to be fine with parallel queries, I believe the Interner must live globally (not per-thread); is that an accurate representation of the code today? Otherwise as far as I can tell a `Symbol` created on thread A can be transferred to other threads, which would lead to problems.

rust-highfive commented:
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/366836995) ([raw log](https://api.travis-ci.org/v3/job/366836995/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
Receiving objects: 100% (102/102), 11.39 KiB | 11.39 MiB/s, done.
---
Resolving deltas: 100% (97/97), completed with 89 local objects.
---
[00:00:45] configure: rust.quiet-tests     := True
---
[00:13:25] 115 |                     ty::BrNamed(id, name)))
[00:13:25]     |                                     ^^^^ expected struct `syntax::symbol::InternedString`, found struct `syntax::symbol::LocalInternedString`
---
[00:13:25] 127 |                     name,
[00:13:25]     |                     ^^^^ expected struct `syntax::symbol::InternedString`, found struct `syntax::symbol::LocalInternedString`
---
[00:13:25] 135 |                     bound_region: ty::BrNamed(id, name)
[00:13:25]     |                                                   ^^^^ expected struct `syntax::symbol::InternedString`, found struct `syntax::symbol::LocalInternedString`
---
[00:13:31] 889 |             name: l.lifetime.name.name().as_str(),
[00:13:31]     |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `syntax::symbol::InternedString`, found struct `syntax::symbol::LocalInternedString`
---
[00:13:31] 1430 |             name: param.lifetime.name.name().as_str(),
[00:13:31]      |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `syntax::symbol::InternedString`, found struct `syntax::symbol::LocalInternedString`
---
[00:13:31]   process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustc --crate-name rustc_typeck librustc_typeck/lib.rs --color always --error-format json --crate-type dylib --emit=dep-info,link -C prefer-dynamic -C opt-level=2 -C metadata=d7a9eeb5d1acb48b -C extra-filename=-d7a9eeb5d1acb48b --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/release/deps --extern fmt_macros=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libfmt_macros-3706e912fdb98df1.so --extern rustc_data_structures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_data_structures-93cb1ddd29ab61a4.so --extern rustc_const_math=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_const_math-895f7ddc4467bb8d.so --extern arena=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libarena-f2778bd6cd1c5bdd.so --extern rustc_platform_intrinsics=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_platform_intrinsics-6f259a2a9f59267f.so --extern rustc=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc-23f5dcecdda8feb4.so --extern rustc_errors=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_errors-609e2421d03f9c9a.so --extern log=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/liblog-3e35efb9e5bc7a9a.rlib --extern syntax=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libsyntax-d3b6fcf798f7d22a.so --extern syntax_pos=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/deps/libsyntax_pos-9f3518d56a01456f.so -L native=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/build/backtrace-sys-8b35e3c2ea935fab/out/.libs -L native=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-rustc/x86_64-unknown-linux-gnu/release/build/miniz-sys-63734d0048644b22/out` (exit code: 101)
---
121760 ./obj/build/bootstrap/debug/incremental/bootstrap-351vorei3hhuv/s-f059e4u5y0-1ocolbt-3ssff2uugm5rt
---
$ ls -lat $HOME/Library/Logs/DiagnosticReports/
ls: cannot access /home/travis/Library/Logs/DiagnosticReports/: No such file or directory
travis_time:end:1322ddb0:start=1523812471589806466,finish=1523812471597800232,duration=7993766
travis_fold:end:after_failure.2
travis_fold:start:after_failure.3
travis_time:start:1346d13a
$ find $HOME/Library/Logs/DiagnosticReports -type f -name '*.crash' -not -name '*.stage2-*.crash' -not -name '
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/366839002) ([raw log](https://api.travis-ci.org/v3/job/366839002/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
Receiving objects: 100% (102/102), 11.41 KiB | 11.41 MiB/s, done.
---
Resolving deltas: 100% (97/97), completed with 89 local objects.
---
[00:00:48] configure: rust.quiet-tests     := True
---
[00:34:20] 227 |                         hir::LifetimeName::Name(Symbol::intern(&p.name))
[00:34:20]     |                                                                ^^^^^^^ expected str, found struct `syntax::symbol::InternedString`
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/366876345) ([raw log](https://api.travis-ci.org/v3/job/366876345/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:00:46] configure: rust.quiet-tests     := True
---
[00:46:11] .................................................................................i..................
[00:46:18] ........................i...........................................................................
---
[00:46:59] .......................i...........................................................................i
[00:47:06] ....................................................................................................
[00:47:12] .............ii.....................................................................................
---
[00:47:58] .............................................i......................................................
---
[00:51:53] ..............................i.....................................................................
[00:52:07] ...............................................................i....................................
[00:52:22] .................................................i..................................................
[00:52:42] ....................................................................................................
[00:53:04] ....................................................................................................
[00:53:25] ....................................................................................................
[00:53:50] .......i............................................................................................
[00:54:18] ...i.................................................................................test [run-pass] run-pass/mir_heavy_promoted.rs has been running for over 60 seconds
[00:54:26] ...............
[00:54:58] ....................................................................................................
[00:55:30] .....................................................................ii.............................
[00:56:21] ................................i....................................................i.ii..test [run-pass] run-pass/saturating-float-casts.rs has been running for over 60 seconds
[00:56:25] .........
[00:57:08] .............................................................................................iiiiiii
---
[00:59:22] ...................i............................................................ii.iii..............
[00:59:29] ....................................................................................................
[00:59:37] ........i..............................i............................................................
[00:59:44] ....................................................................................................
[00:59:51] ..........i.........................................................................................
[00:59:59] ....................................................................................................
[01:00:09] ....................................................................................................
[01:00:19] ....................................................................................................
[01:00:29] ....................................................................................................
[01:00:43] ....................................................................................................
[01:00:51] ...i................................................................................................
[01:01:00] .......i..ii........................................................................................
[01:01:10] ....................................................................................................
[01:01:19] ....................................................................................................
[01:01:29] .........................................................................i..........................
[01:01:39] ...................i................................................................................
---
[01:02:13] ...........................i........................................................................
[01:02:14] ....................................................................i...............................
[01:02:16] ................i.......................................................
---
[01:02:30] .............i........................
---
[01:03:00] i...i..ii....i.............ii.........iii......i..i...i...ii..i..i..ii.....
---
[01:03:03] i.......i......................i.......
---
[01:03:41] iiii.......i..i........i..i.i.............i..........iiii...........i...i..........ii.i.i.......ii..
[01:03:42] ....ii...
---
[01:11:33] ..................i.................................................................................
[01:13:09] .......i............................................................................................
---
[01:15:13] .....................................i..............................................................
[01:15:33] ....................................................................................................
[01:15:53] ..........................................i.........................................................
---
[01:17:27] .........................................................ii.........................................
---
[01:18:47] .....................i..............................................................................
---
[01:24:06] ii..................................................................................................
[01:24:24] ....................................................................................................
[01:24:41] .......................iii......i......i...i......i.................................................
[01:24:50] ....................................................................................................
[01:25:05] .............................................iiii........ii.........................................
[01:25:16] ....................................................................................................
[01:25:32] ............................................................................................i.......
[01:25:56] ....................................................................................................
[01:26:07] ....................................................................................................
[01:26:17] ..iiii...............................................
---
[01:29:20] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "--release" "--locked" "--color" "always" "--features" " jemalloc" "--manifest-path" "/checkout/src/rustc/Cargo.toml" "-p" "rustc_driver" "--" "--quiet"
[01:29:20] expected success, got: exit code: 101
[01:29:20]
[01:29:20]
[01:29:20] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:29:20] Build completed unsuccessfully in 0:44:25
[01:29:20] make: *** [check] Error 1
[01:29:20] Makefile:58: recipe for target 'check' failed
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Zoxc commented:
cc @rust-lang/compiler - Just wondering if there are any objections to this

bors commented:
:sunny: Test successful - [status-travis](https://travis-ci.org/rust-lang/rust/builds/369131422?utm_source=github_status&utm_medium=notification)
State: approved= try=True

michaelwoerister commented:
@Mark-Simulacrum, would you mind starting a perf run for this?

Link: http://perf.rust-lang.org/compare.html?start=85f5dd489ea1cb93707ba1ddf4a7678d3aee3f74&end=7ef4d38de555741a0008ab09b945d77a2e9c0b65&stat=instructions%3Au

Mark-Simulacrum commented:
Perf queued.

michaelwoerister commented:
OK, performance looks good! r=me after rebase.

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/371225480) ([raw log](https://api.travis-ci.org/v3/job/371225480/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
/usr/local/lib/python2.7/dist-packages/pip/_vendor/requests/packages/urllib3/util/ssl_.py:122: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. You can upgrade to a newer version of Python to solve this. For more information, see https://urllib3.readthedocs.io/en/latest/security.html#insecureplatformwarning.
  InsecurePlatformWarning
/usr/local/lib/python2.7/dist-packages/pip/_vendor/requests/packages/urllib3/util/ssl_.py:122: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. You can upgrade to a newer version of Python to solve this. For more information, see https://urllib3.readthedocs.io/en/latest/security.html#insecureplatformwarning.
  InsecurePlatformWarning
  Downloading https://files.pythonhosted.org/packages/4a/7d/9c0f35dc594b78137929523f9632ec649b69af2d30ca6a7a8c60d414b23a/awscli-1.15.9-py2.py3-none-any.whl (1.3MB)
    0% |▎                               | 10kB 11.2MB/s eta 0:00:01
    1% |▌                               | 20kB 1.8MB/s eta 0:00:01
    2% |▉                               | 30kB 2.1MB/s eta 0:00:01
    3% |█                               | 40kB 2.0MB/s eta 0:00:01
---
    100% |████████████████████████████████| 61kB 7.3MB/s 
Collecting botocore==1.10.9 (from awscli)
/usr/local/lib/python2.7/dist-packages/pip/_vendor/requests/packages/urllib3/util/ssl_.py:122: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. You can upgrade to a newer version of Python to solve this. For more information, see https://urllib3.readthedocs.io/en/latest/security.html#insecureplatformwarning.
  InsecurePlatformWarning
  Downloading https://files.pythonhosted.org/packages/fe/13/45a8eeb5d78e2578b8e55df58e3921369efe51eaa57a9d2a36e7bef45bcc/botocore-1.10.9-py2.py3-none-any.whl (4.2MB)
    0% |                                | 10kB 31.5MB/s eta 0:00:01
    0% |▏                               | 20kB 33.6MB/s eta 0:00:01
    0% |▎                               | 30kB 40.8MB/s eta 0:00:01
    0% |▎                               | 40kB 17.9MB/s eta 0:00:01
---

[00:04:27] travis_fold:start:tidy
travis_time:start:tidy
tidy check
[00:04:28] tidy error: /checkout/src/librustc/hir/map/def_collector.rs:227: line longer than 100 chars
[00:04:29] some tidy checks failed
[00:04:29] 
[00:04:29] 
[00:04:29] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor" "--quiet"
[00:04:29] 
[00:04:29] 
[00:04:29] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:04:29] Build completed unsuccessfully in 0:01:48
[00:04:29] Build completed unsuccessfully in 0:01:48
[00:04:29] Makefile:79: recipe for target 'tidy' failed
[00:04:29] make: *** [tidy] Error 1

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:292ac28c
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Zoxc commented:
@bors r=michaelwoerister

bors commented:
:pushpin: Commit adc542b has been approved by `michaelwoerister`

<!-- @bors r=michaelwoerister adc542bb25fea74f22d4fe49953847d2b7396620 -->

bors commented:
:umbrella: The latest upstream changes (presumably #50016) made this pull request unmergeable. Please resolve the merge conflicts.

Zoxc commented:
@bors r=michaelwoerister

bors commented:
:pushpin: Commit e591439 has been approved by `michaelwoerister`

<!-- @bors r=michaelwoerister e5914391e41bd939c9ff798a80c5cb068ed70a0b -->

bors commented:
:umbrella: The latest upstream changes (presumably #50228) made this pull request unmergeable. Please resolve the merge conflicts.

Zoxc commented:
@bors r=michaelwoerister

bors commented:
:pushpin: Commit ce387b8 has been approved by `michaelwoerister`

<!-- @bors r=michaelwoerister ce387b81381c8338a195befd13e3b92015e4fee0 -->

Zoxc commented:
@bors r=michaelwoerister

bors commented:
:pushpin: Commit 4d52751 has been approved by `michaelwoerister`

<!-- @bors r=michaelwoerister 4d52751d12429f4508f9a3c537521ba1113fabf5 -->

eddyb commented on src/libsyntax_pos/symbol.rs:
```diff
@@ -365,84 +377,208 @@ fn with_interner<T, F: FnOnce(&mut Interner) -> T>(f: F) -> T {
+/// Represents a string stored in the string interner
+#[derive(Clone, Copy, Eq)]
+pub struct InternedString {
+    symbol: Symbol,
```
I saw this PR too late, but wish this was named `SymbolContents` instead of `InternedString`, since AFAICT all it does is it changes the behavior of comparison / hashing / etc. to act on the string value inside.

eddyb replied:
`SymbolAsStr`? I mean, I want the type to evoke "this is a `Symbol` made to look like the string contents thereof".
